Partitions: Include ACL Info and API cross-ref for CLI Commands

jkirschner-hashicorp opened the pull request:
## Description

This draft PR serves two purposes:
- Close #12035
- Improve the structure of info introduced by #11984
  - Align on a new structure for this information within the admin partition command section that can be applied to all other commands
  - This is a follow-up to the long conversation @karl-cardenas-coding and @trujillo-adam started on a previous draft PR: https://github.com/hashicorp/consul/pull/11031#issuecomment-938952197
  
## Previews

[Partition command page](https://consul-mikdi9uno-hashicorp.vercel.app/commands/partition)

![image](https://user-images.githubusercontent.com/85913323/157716721-65d3e909-5ad8-4492-9a4d-3e9dbf59bc29.png)

[Partition HTTP API page](https://consul-mikdi9uno-hashicorp.vercel.app/api-docs/admin-partitions)

![image](https://user-images.githubusercontent.com/85913323/157716645-ea81ca43-1e7f-4322-97f3-9a1c789ce8b0.png)

jkirschner-hashicorp committed:
docs: add partition command characteristics

Characteristics include:
- Required ACL permissions
- Corresponding HTTP API endpoint
- (Lack of) support for blocking queries and agent caching

jkirschner-hashicorp commented:
@karl-cardenas-coding , @trujillo-adam : would appreciate your perspective on the different options for showing required ACLs, corresponding HTTP API, etc for CLI commands. [This page](https://consul-d0mainfxr-hashicorp.vercel.app/commands/partition) has a different approach shown for each of the 5 subcommands.

#11984 introduced a less-than-ideal way to include this info which we should improve. I'm trying to use this PR (limited just to partition commands) to align on that improvement.

@trujillo-adam : one thing I noticed is that the current partition command page does things a bit differently than the other command pages. For example, it has one page for all commands/subcommands (rather than a separate page for create, read, update, delete, and list). It also structures the content within a command section different (e.g., one continuous section rather than breaking it up with an `API Options` subsection ([like on this page](https://www.consul.io/commands/operator/area#create)). Does the partition page show the new style we want to apply (eventually) everywhere else?

karl-cardenas-coding commented:
@jkirschner-hashicorp I really like how the `list` command example is looking. It appears less busy and the table really helps the reader focus the attention on the information. It felt less confusing trying to figure what was happening when the table form was used.

jkirschner-hashicorp commented:
@karl-cardenas-coding: I think the `list` option is also my favorite by far. I'm not sold on the table heading names. Unfortunately, that component doesn't allow a table without headers (and having an empty first row looks weird / like there's a bug)... so we need _some_ header labels.

Could "Name" and "Value" be better than "Detail" and "Info" (which feel like synonyms)?

I'll wait until @trujillo-adam weighs in as well before proceeding.

jkirschner-hashicorp commented:
@karl-cardenas-coding : per [the style guide](https://github.com/hashicorp/learn/blob/master/STYLE_GUIDE.md#tables), @trujillo-adam recommended the list approach rather than the table approach.

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
```diff
@@ -66,6 +66,13 @@ You can issue the following subcommands with the `consul partition` command.
 
 The `create` subcommand sends a request to the server to create a new admin partition.
+This subcommand has the following characteristics:
+- **[Required ACLs](/docs/security/acl/acl-system)**: `operator:write`
```
I decided to link to the ACL docs in case someone reading this doesn't have ACLs enabled and isn't super familiar with them. Seem like a good idea?

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
```diff
@@ -95,7 +102,17 @@ $ consul partition create -name "webdev" -description "Partition for admin of we
 
 ### `write`
 
-The `write` subcommand sends a request to the server to create a new admin partition or update an existing partition from its full definition. You can specify an admin partition definition file or use values from `stdin`.
```
I moved
> You can specify an admin partition definition file or use values from `stdin`.

to the usage section just below

jkirschner-hashicorp replied:
No longer relevant - without the usage subsection, it's no longer necessary to move this part

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
````diff
@@ -109,7 +126,7 @@ Use the following syntax to write from `stdin`:
 consul partition write <OPTIONS> -
 ```
 
-The definition file or `stdin` values can be provided in JSON or HCL format. Refer to the [Admin Partition Definition](#partition-definition) section for details about the supported parameters.
````
I moved the part about JSON or HCL format up (see L115)

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
```diff
@@ -134,6 +151,14 @@ $ consul partition write -format json -show-meta - <<< 'name = "webdev-bu" descr
 The `read` subcommand sends a request to the server to read the configuration for the specified partition and print it to the console.
+This subcommand has the following characteristics:
+- **[Required ACLs](/docs/security/acl/acl-system)**: `operator:read`
+  - `none` is sufficient to allow a non-anonymous token to read its own partition
```
I wasn't sure the best way to deal with this nuance.

The [HTTP API docs](https://www.consul.io/api-docs/admin-partitions#read-a-partition) handle this by using a footnote on `none` (see screenshot below). I'm not sure if a footnote approach like that would look strange below a list. I also wanted to see if something like this would be clearer.

![image](https://user-images.githubusercontent.com/85913323/155604457-e2ab1440-3fee-4520-b264-51179a79b548.png)

jkirschner-hashicorp replied:
I've tried it with and without an asterisk.

jkirschner-hashicorp committed:
docs: restructure partition API characteristics

The existing characteristics were restructured into a list.
The corresponding CLI command characteristic was added.

jkirschner-hashicorp commented:
@trujillo-adam : I believe I've addressed all the feedback we discussed on Slack. Ready for final review.

Once this is approved, I can extend to all non-partition commands and HTTP API endpoints.

jkirschner-hashicorp commented:
@trujillo-adam : I've updated this to use the two-column table approach. Ready for review.

jkirschner-hashicorp commented on website/content/api-docs/admin-partitions.mdx:
```diff
@@ -13,23 +13,18 @@ The functionality described here is available only in
+This endpoint creates a new Partition and has the following characteristics:
+
+| Characteristic | Value |
+| :------------- | :---- |
```
The colon forces left-aligned header text

jkirschner-hashicorp replied:
Yes, we will maintain the flexibility to change alignment. We just want to make the default alignment be correct most of the time (left-aligned). This is captured in more detail in Asana: https://app.asana.com/0/1100423001970639/1201999608008890/f

jkirschner-hashicorp commented on website/content/api-docs/admin-partitions.mdx:
```diff
@@ -175,26 +158,18 @@ deleted all the associated partitioned data in the background. Only once
+| :------------- | :---- |
+| [HTTP Method](/api-docs#http-methods) | `DELETE` |
+| [URL Path](/api-docs#version-prefix) | `/partition/:name` |
+| [Response Type](/api-docs#formatted-json-output) | none |
```
Choose to put "none" rather than "N/A" for the value here because, technically, a DELETE request _can_ respond with what was deleted, we just choose not to in Consul.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12045)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12045)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=12045).</sub>

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12045)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=12045)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=12045).</sub>

trujillo-adam reviewed (commented):
I left a few minor suggestions and comments. Some of the content outside the scope of this PR needs to be edited, e.g., the description of the delete partition endpoint, but I'll open a new PR to fix those.

jkirschner-hashicorp committed:
docs: add partial for api/cli characteristics links

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
```diff
@@ -66,6 +73,16 @@ You can issue the following subcommands with the `consul partition` command.
+| [Blocking Queries](/api-docs/features/blocking) | not supported from commands, but may be from the corresponding HTTP API endpoint |
+| [Agent Caching](/api-docs/features/caching) | not supported from commands, but may be from the corresponding HTTP API endpoint |
+
+#### Usage
```
I added the "Usage" subheading because I thought the content looked strange otherwise. It just jumps from the characteristics table to a shell command box. See image below showing what it looks like with and without that heading. Let me know how you want me to proceed.

![image](https://user-images.githubusercontent.com/85913323/159790559-892e5e76-33ff-4aed-bce7-a73a0a930bed.png)

jkirschner-hashicorp replied:
@trujillo-adam as we discussed, I removed both the usage subheading **and** the usage command string. [Preview here](https://consul-602w17yc3-hashicorp.vercel.app/commands/partition).

It now looks like:
![image](https://user-images.githubusercontent.com/85913323/162082565-c7a5fc04-1453-47b0-aa3a-2c3ae9823c5f.png)

jkirschner-hashicorp commented:
@trujillo-adam : ready for final review. I think the last call to make here is whether or not to have "Usage" subheaders, per this comment: https://github.com/hashicorp/consul/pull/12045#discussion_r833697767

jkirschner-hashicorp committed:
docs: remove partition subcommand usage headings

jkirschner-hashicorp commented on website/content/commands/partition.mdx:
````diff
@@ -109,7 +123,7 @@ Use the following syntax to write from `stdin`:
 ```
 
-The definition file or `stdin` values can be provided in JSON or HCL format. Refer to the [Admin Partition Definition](#partition-definition) section for details about the supported parameters.
+The definition file or `stdin` values can be provided in JSON or HCL format. Refer to the [Admin Partition Definition](#admin-partition-definition) section for details about the supported parameters.
````
Fixed a broken anchor link

jkirschner-hashicorp commented:
@karl-cardenas-coding, @trujillo-adam : ready for final review/approval, all comments addressed. See previews:
- API docs page: https://consul-dl3fncz9o-hashicorp.vercel.app/api-docs/admin-partitions
- Command docs page: https://consul-dl3fncz9o-hashicorp.vercel.app/commands/partition

trujillo-adam reviewed (commented):
One more suggestion to try out

jkirschner-hashicorp committed:
docs: adjust HTTP API/CLI characteristics tables

jkirschner-hashicorp commented:
@trujillo-adam : based on our latest conversation, I've updated the PR.

### HTTP API endpoint change
A long time has passed since we started on this. To make sure you still think this is a good change to make, I wanted to include this before/after comparison for an HTTP API endpoint for you to consider.

![ComparisonOfCharacteristicsTable](https://user-images.githubusercontent.com/85913323/180096661-e0165c5f-35bf-4124-9e4a-0239c05a804d.PNG)

### Command change

**Question:** Do you think mentioning blocking queries and agent caching on every CLI command is valuable? The text is the same on all of the commands: no command supports these things, though their corresponding HTTP API endpoints _might_. I'm on the fence about whether it's better to:
- Exclude this information as unnecessary noise from the command docs, assuming that this isn't a valuable place to expose them to the existence of these features
- Expose the reader to these features, even if only available from the HTTP API, and knowing that this content will only be relevant temporarily (first exposure) and then noise afterwards

I'm leaning towards removing those two rows, but can be convinced either way. Thoughts?

![image](https://user-images.githubusercontent.com/85913323/180097060-db855d6d-e99c-49f5-ad1d-492cc55392f5.png)

jkirschner-hashicorp commented:
Latest preview links:
- [Partition command page](https://consul-mikdi9uno-hashicorp.vercel.app/commands/partition)
- [Partition HTTP API page](https://consul-mikdi9uno-hashicorp.vercel.app/api-docs/admin-partitions)

jkirschner-hashicorp committed:
docs: remove unnecessary partition CLI cmd info
